org.eclipse.cdt.cross.arc.gnu/src/org/eclipse/cdt/cross/arc/gnu/common/CompilerPluginVersionsChecker.java

```diff
@@ -0,0 +1,103 @@
+import org.eclipse.swt.widgets.Display;
+import org.osgi.service.prefs.BackingStoreException;
+
+public class CompilerPluginVersionsChecker {
+
+    private final String FILENAME_WITH_PROJECT_SCOPED_PREFERENCES = "com.arc.cdt";
+    private final String UCLIBC_PLUGIN_VERSION = "arc.uclibc_plugin.version";
+    private final String ELF_PLUGIN_VERSION = "arc.elf32_plugin.version";
+
+    // These version numbers should be incremented when incompatible changes appear in these
+    // plug-ins.
+    private final int UCLIBC_PLUGIN_VERSION_NUMBER = 1;
+    private final int ELF_PLUGIN_VERSION_NUMBER = 1;
+
+    private final int UNREAL_VERSION_NUMBER = -1;
+
+    private HashSet<IProject> seenProjects = new HashSet<>();
+
+    private static CompilerPluginVersionsChecker INSTANCE = null;
+
+    private CompilerPluginVersionsChecker() {
+    }
+
+    public static CompilerPluginVersionsChecker getCompilerPluginVersionsChecker() {
+        if (INSTANCE == null) {
+            INSTANCE = new CompilerPluginVersionsChecker();
+        }
+        return INSTANCE;
+    }
+
+    void checkPluginVersions(IScannerInfoCollector collector) {
+        // Write plug-in versions to .settings/com.arc.cdt file if they are not
+        // present, check them otherwise and show warning if there is a
+        // mismatch.
+        InfoContext oC;
+        if ((collector instanceof PerProjectSICollector)) {
+            oC = ((PerProjectSICollector) collector).getContext();
+            IProject oProject = oC.getProject();
+            if (seenProjects.contains(oProject)) {
+                return;
+            }
+            ProjectScope projectScope = new ProjectScope(oProject);
+            IEclipsePreferences pref = projectScope
+                    .getNode(FILENAME_WITH_PROJECT_SCOPED_PREFERENCES);
+            int projectElfPluginVersionNumber = pref.getInt(ELF_PLUGIN_VERSION,
+                    UNREAL_VERSION_NUMBER);
+            int projectUclibcPluginVersionNumber = pref.getInt(UCLIBC_PLUGIN_VERSION,
+                    UNREAL_VERSION_NUMBER);
+            String warningMsg = "";
+            String warningMsgPattern = "Your %s plug-in version is %d, but the project %s was created with the"
+                    + " plug-in version %d.\n";
+            if (projectElfPluginVersionNumber != UNREAL_VERSION_NUMBER) {
+                if (projectElfPluginVersionNumber != ELF_PLUGIN_VERSION_NUMBER) {
+                    warningMsg += String.format(warningMsgPattern, "elf", ELF_PLUGIN_VERSION_NUMBER,
+                            oProject.getName(), projectElfPluginVersionNumber);
+                }
+                if (projectUclibcPluginVersionNumber != UCLIBC_PLUGIN_VERSION_NUMBER) {
+                    warningMsg += String.format(warningMsgPattern, "uclibc",
+                            UCLIBC_PLUGIN_VERSION_NUMBER, oProject.getName(),
+                            projectUclibcPluginVersionNumber);
+                }
+            }
```

Comment:
Like in debugger, it shouldn't accept projects that were created when version numbers were not tracked at all.